malariagen_data/anopheles.py

```diff
@@ -7726,6 +7727,550 @@ def plot_h1x_gwss(
+            # N.B., do not be tempted to convert this sample query into integer
+            # indices using _prep_sample_selection_params, because the indices
+            # are different in the haplotype data.
+            sample_query=sample_query,
+            cohort_size=cohort_size,
+            random_seed=random_seed,
+        )
+
+        try:
+            results = self.results_cache_get(name=name, params=params)
+
+        except CacheMiss:
+            results = self._ihs_gwss(**params)
+            self.results_cache_set(name=name, params=params, results=results)
+
+        x = results["x"]
+        ihs = results["ihs"]
+
+        return x, ihs
+
+    def _ihs_gwss(
+        self,
+        contig,
+        analysis,
+        sample_sets,
+        sample_query,
+        window_size,
+        window_percentiles,
+        standardize,
+        standardization_bins,
+        standardization_n_bins,
+        standardization_diagnostics,
+        filter_min_maf,
+        compute_min_maf,
+        min_ehh,
+        max_gap,
+        gap_scale,
+        include_edges,
+        use_threads,
+        cohort_size,
+        random_seed,
+    ):
+        ds_haps = self.haplotypes(
+            region=contig,
+            analysis=analysis,
+            sample_query=sample_query,
+            sample_sets=sample_sets,
+            cohort_size=cohort_size,
+            random_seed=random_seed,
+        )
+
+        gt = allel.GenotypeDaskArray(ds_haps["call_genotype"].data)
+        with self._dask_progress(desc="Load haplotypes"):
+            ht = gt.to_haplotypes().compute()
+
+        pos = ds_haps["variant_position"].values
+
+        if filter_min_maf > 0:
+            ac = ht.count_alleles().to_frequencies()
+            maf_filter = ac[:, 1] > filter_min_maf
+            ht = ht.compress(maf_filter, axis=0)
+            pos = pos[maf_filter]
+
+        ihs = allel.ihs(
+            h=ht,
+            pos=pos,
+            min_maf=compute_min_maf,
+            min_ehh=min_ehh,
+            include_edges=include_edges,
+            max_gap=max_gap,
+            gap_scale=gap_scale,
+            use_threads=use_threads,
+        )
+
+        if standardize:
+            ihs = allel.standardize_by_allele_count(
+                score=ihs,
+                aac=ht.count_alleles()[:, 1],
+                bins=standardization_bins,
+                n_bins=standardization_n_bins,
+                diagnostics=standardization_diagnostics,
+            )
+            ihs = ihs[0]
+
+        if window_size:
+            # remove any NaNs prior to windowed percentiles
+            na_mask = np.isnan(ihs)
+            ihs = ihs[~na_mask]
+            pos = pos[~na_mask]
+
+            ihs = allel.moving_statistic(
+                ihs, statistic=np.percentile, size=window_size, q=window_percentiles
+            )
+            pos = allel.moving_statistic(pos, statistic=np.mean, size=window_size)
```

Comment:
Couple of things here.

In the common case of default parameters, we'll need the allele counts twice, once for the minor allele frequency filtering, and again for the standardisation. Would be better to compute allele counts once up front, then re-use for both.

Also, would probably be better to remove nans before standardisation. Not sure it matters much, but might be a little more efficient if nothing else.

```suggestion

        # compute allele counts up front
        ac = ht.count_alleles(max_allele=1)
        
        # filter haplotypes by minor allele frequency
        if filter_min_maf > 0:
            af = ac.to_frequencies()
            # compute minor allele frequency, N.B., could be either ref or alt allele
            maf = np.min(af, axis=1)
            maf_filter = maf > filter_min_maf
            ht = ht.compress(maf_filter, axis=0)
            pos = pos[maf_filter]
            ac = ac[maf_filter]

        # compute IHS
        ihs = allel.ihs(
            h=ht,
            pos=pos,
            min_maf=compute_min_maf,
            min_ehh=min_ehh,
            include_edges=include_edges,
            max_gap=max_gap,
            gap_scale=gap_scale,
            use_threads=use_threads,
        )

        # remove any NaN values
        na_mask = ~np.isnan(ihs)
        ihs = ihs[na_mask]
        pos = pos[na_mask]
        ac = ac[na_mask]
        
        if standardize:
            # standardize IHS values
            ihs, _ = allel.standardize_by_allele_count(
                score=ihs,
                aac=ac[:, 1],
                bins=standardization_bins,
                n_bins=standardization_n_bins,
                diagnostics=standardization_diagnostics,
            )

        if window_size:
            # summarize IHS values in windows
            ihs = allel.moving_statistic(
                ihs, statistic=np.percentile, size=window_size, q=window_percentiles
            )
            pos = allel.moving_statistic(pos, statistic=np.mean, size=window_size)
```

Reply:
Also minor allele frequency calculation needs to account for minor allele being either ref or alt.